src/parse/asp/builtins.go

```diff
@@ -753,6 +754,62 @@ func addDep(s *scope, args []pyObject) pyObject {
 	return None
 }
 
+func addDatumToTargetAndMaybeQueue(s *scope, target *core.BuildTarget, datum core.BuildInput, systemAllowed, tool bool) {
+	target.AddDatum(datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+func addNamedDatumToTargetAndMaybeQueue(s *scope, name string, target *core.BuildTarget, datum core.BuildInput, systemAllowed, tool bool) {
+	target.AddNamedDatum(name, datum)
+	// Queue this dependency if it'll be needed.
+	if l, ok := datum.Label(); ok && target.State() > core.Inactive {
+		err := s.state.QueueTarget(l, target.Label, true, false)
+		s.Assert(err == nil, "%s", err)
+	}
+}
+
+// Add runtime dependencies to target
+func addData(s *scope, args []pyObject) pyObject {
+	s.Assert(s.Callback, "can only be called from a pre- or post-build callback")
+
+	label := args[0]
```

Comment:
super nit: might do a `datum := args[1]` here since you repeatedly use it below (and IMO it just reads a bit nicer to name them)